jinahub/indexers/searcher/HnswlibSearcher/hnswlib_searcher.py

```diff
@@ -1,128 +1,271 @@
+            self._init_empty_index()
+
+        self._index.set_ef(self.ef_query)
 
-    def _load_index(self, ids, vecs):
-        for idx, v in enumerate(vecs):
-            self._indexer.add_items(v.astype(np.float32), idx)
-            self._doc_id_to_offset[ids[idx]] = idx
-        self._indexer.set_ef(self.ef_query)
+    def _init_empty_index(self):
+        self._index.init_index(
+            max_elements=self.max_elements,
+            ef_construction=self.ef_construction,
+            M=self.max_connection,
+        )
+        self._ids_to_inds = bidict()
 
     @requests(on='/search')
-    def search(self, docs: Optional[DocumentArray], parameters: Dict, **kwargs):
+    def search(
+        self, docs: Optional[DocumentArray] = None, parameters: Dict = {}, **kwargs
+    ):
+        """Attach matches to the Documents in `docs`, each match containing only the
+        `id` of the matched document and the `score`.
+
+        :param docs: An array of `Documents` that should have the `embedding` property
+            of the same dimension as vectors in the index
+        :param parameters: Dictionary with optional parameters that can be used to
+            override the parameters set at initialization. Supported keys are
+            `traversal_paths`, `top_k` and `ef_query`.
         """
+        if docs is None:
+            return
 
-        :param docs: `Document` with `.embedding` the same shape as the `Documents`
-            it has stored.
-        :param parameters: dictionary to define the `traversal_paths` and the `top_k`.
+        traversal_paths = parameters.get('traversal_paths', self.traversal_paths)
+        docs_search = docs.traverse_flat(traversal_paths)
+        if len(docs_search) == 0:
+            return
+
+        ef_query = parameters.get('ef_query', self.ef_query)
+        top_k = parameters.get('top_k', self.top_k)
+
+        self._index.set_ef(ef_query)
+
+        if top_k > len(self._ids_to_inds):
+            self.logger.warning(
+                f'The `top_k` parameter is set to a value ({top_k}) that is higher than'
+                f' the number of documents in the index ({len(self._ids_to_inds)})'
+            )
+            top_k = len(self._ids_to_inds)
+
+        embeddings_search = docs_search.embeddings
+        if embeddings_search.shape[1] != self.dim:
+            raise ValueError(
+                'Query documents have embeddings with dimension'
+                f' {embeddings_search.shape[1]}, which does not match the dimension of'
+                f' the index ({self.dim})'
+            )
 
-        :return: Attaches matches to the Documents sent as inputs, with the id of the
-            match, and its embedding.
+        indices, dists = self._index.knn_query(docs_search.embeddings, k=top_k)
+
+        for i, (indices_i, dists_i) in enumerate(zip(indices, dists)):
+            for idx, dist in zip(indices_i, dists_i):
+                match = Document(id=self._ids_to_inds.inverse[idx])
```

Comment:
Should also add the embedding, or am I missing something?

Reply:
You are not missing anything - I have conciously decided against it.

Filling the document with attributes is the responsibility of the storage indexer. The search indexer is not even guaranteed to have the embedding (e.g. PQ Faiss), and may not be optimized for its retrieval even if it has it.

Reply:
Can you make a note about that in the README? It will require the user to add another parameter to the search request in order to get the embedding from PSQL (see https://github.com/jina-ai/executors/blob/main/jinahub/indexers/storage/PostgreSQLStorage/postgres_indexer.py#L208)

Reply:
Done